File: src/engine/linear-portable.ts
/**
 * Runtime evaluator for the portable spline+logistic resolvers exported by
 * `analysis/27_export_portable.py` (format `linear-portable-1`).
 *
 * TS port of `analysis/lib_py/linear_portable.py::predict_proba_linear` — keep
 * the two behaviourally identical. Each numeric feature contributes to the
 * decision function via either an exact per-segment cubic (a SplineTransformer
 * feature; constant past the ends) or a plain linear term (a StandardScaler
 * feature). Categoricals are one-hot weight rows; an unknown category
 * contributes zero (`handle_unknown='ignore'`).
 */

export interface LinearCubicNumeric {
  kind: "pw_cubic";
  /** segment breakpoints (spline interior knots); `seg.length === breaks.length - 1` */
  breaks: number[];
  /** `seg[s][k]` = `[a, b, c, d]`, contribution `= a + b·u + c·u² + d·u³`, `u = x - breaks[s]` */
  seg: number[][][];
}

export interface LinearScalarNumeric {
  kind: "linear";
  mean: number;
  scale: number;
  /** per effective class */
  w: number[];
}

export type LinearNumeric = LinearCubicNumeric | LinearScalarNumeric;

export interface LinearPortableModel {
  format: "linear-portable-1";
  objective: "binary" | "multiclass";
  /** raw-prediction class order (LogisticRegression.classes_) */
  classes: string[];
  /** semantic label order to key results by */
  labels: string[];
  feature_names: string[];
  /** per effective class (1 for binary, n_classes for multinomial) */
  intercept: number[];
  numeric: Record<string, LinearNumeric>;
  /** `categorical[name][value]` = per-class weights; `_unknown` = zeros */
  categorical: Record<string, Record<string, number[]>>;
}

export type FeatureVec = Record<string, number | string | null | undefined>;

function numericContrib(entry: LinearNumeric, x: number, k: number): number {
  if (entry.kind === "linear") {
    return ((x - entry.mean) / entry.scale) * (entry.w[k] ?? 0);
  }
  const { breaks, seg } = entry;
  const first = breaks[0] ?? 0;
  const last = breaks[breaks.length - 1] ?? 0;
  let s: number;
  let u: number;
  if (x <= first) {
    s = 0;
    u = 0;
  } else if (x >= last) {
    s = seg.length - 1;
    u = last - (breaks[breaks.length - 2] ?? last);
  } else {
    s = 0;
    while (s + 1 < breaks.length - 1 && x >= (breaks[s + 1] ?? Infinity)) s++;
    u = x - (breaks[s] ?? 0);
  }
  const [a = 0, b = 0, c = 0, d = 0] = seg[s]?.[k] ?? [];
  return a + b * u + c * u * u + d * u * u * u;
}

/**
 * Evaluate a portable linear model. Returns `{ label: probability }` keyed by
 * `model.labels`. Categorical values are compared as strings; an unknown
 * category contributes zero. Numeric features are assumed present (the callers
 * always supply them).
 */
export function predictProbaLinear(
  model: LinearPortableModel,
  x: FeatureVec,
): Record<string, number> {
  const nEff = model.intercept.length;
  const dec = [...model.intercept];

  for (const [name, entry] of Object.entries(model.numeric)) {
    const v = x[name];
    if (v === null || v === undefined) continue;
    const num = Number(v);
    if (Number.isNaN(num)) continue;
    for (let k = 0; k < nEff; k++) dec[k] = (dec[k] ?? 0) + numericContrib(entry, num, k);
  }

  for (const [name, table] of Object.entries(model.categorical)) {
    const row = table[String(x[name])] ?? table._unknown ?? [];
    for (let k = 0; k < nEff; k++) dec[k] = (dec[k] ?? 0) + (row[k] ?? 0);
  }

  const { classes, labels } = model;
  const byClass: Record<string, number> = {};
  if (model.objective === "binary") {
    const p1 = 1 / (1 + Math.exp(-(dec[0] ?? 0)));
    byClass[classes[0] ?? "0"] = 1 - p1;
    byClass[classes[1] ?? "1"] = p1;
  } else {
    const m = Math.max(...dec);
    const ex = dec.map((d) => Math.exp(d - m));
    const sum = ex.reduce((p, q) => p + q, 0);
    for (let i = 0; i < classes.length; i++) {
      byClass[classes[i] as string] = (ex[i] ?? 0) / sum;
    }
  }
  return Object.fromEntries(labels.map((l) => [l, byClass[l] ?? 0]));
}

File: analysis/lib_py/linear_portable.py
"""Portable export + reference evaluator for the spline+logistic resolvers.

M04/M15/M20/M21/M25a/M25b are sklearn `Pipeline([('pre', ColumnTransformer(
SplineTransformer | StandardScaler | OneHotEncoder)), ('clf', LogisticRegression)])`.

The whole thing is linear in the transformed design matrix, so each *numeric*
feature's contribution to the decision function `Σ_j coef_j · basis_j(x)` is a
1-D function of that feature alone:
  - a SplineTransformer feature -> a piecewise cubic with breakpoints at the
    spline's interior knots (degree 3), constant outside (matching sklearn's
    `extrapolation='constant'`). Exported as exact per-segment cubic coeffs.
  - a StandardScaler feature -> exactly linear: `((x - mean) / scale) * w`.
Categoricals become per-class one-hot weight rows (`handle_unknown='ignore'`
-> unknown category contributes zero).

`predict_proba_linear` then needs no sklearn / B-spline maths. Verified against
`sklearn.predict_proba` to < 1e-6 in `analysis/27_export_portable.py`.
"""

from __future__ import annotations

import math
from typing import Any

import numpy as np


def _blocks(pre: Any):
    return [(name, trans, list(cols)) for name, trans, cols in pre.transformers_ if trans != "drop"]


def _spline_segments(spline: Any, j: int, coef_block: np.ndarray) -> dict:
    """coef_block: (n_eff, n_bases) slice of clf.coef_ for feature j's spline cols.
    Returns exact per-segment cubic coefficients of `Σ_b coef[:,b]·basis_b(x)`."""
    bs = spline.bsplines_[j]
    t, k = bs.t, bs.k                              # knots, degree (3)
    breaks = [float(v) for v in np.unique(t[k:-k])]   # interior knot breakpoints
    n_eff = coef_block.shape[0]

    def contrib(xs: np.ndarray) -> np.ndarray:
        X = np.zeros((len(xs), spline.n_features_in_))
        X[:, j] = xs
        per = spline.transform(np.zeros((1, spline.n_features_in_))).shape[1] // spline.n_features_in_
        basis = spline.transform(X)[:, j * per:(j + 1) * per]      # (len, per)
        return basis @ coef_block.T                                # (len, n_eff)

    seg = []
    for s in range(len(breaks) - 1):
        lo, hi = breaks[s], breaks[s + 1]
        # 4 sample points strictly inside -> exact cubic (the contribution IS cubic here)
        xs = lo + (hi - lo) * np.array([0.12, 0.38, 0.62, 0.88])
        u = xs - lo
        V = np.vstack([np.ones_like(u), u, u ** 2, u ** 3]).T      # (4, 4)
        vals = contrib(xs)                                          # (4, n_eff)
        coeffs = np.linalg.solve(V, vals)                          # (4, n_eff): [a,b,c,d] per class
        seg.append([[float(coeffs[p, kk]) for p in range(4)] for kk in range(n_eff)])
    return {"kind": "pw_cubic", "breaks": breaks, "seg": seg}


def export_linear(pipeline: Any, feature_names: list[str], labels: list[str]) -> dict:
    pre = pipeline.named_steps["pre"]
    clf = pipeline.named_steps["clf"]
    classes = [str(c) for c in clf.classes_]
    n_eff = clf.coef_.shape[0]
    objective = "binary" if n_eff == 1 else "multiclass"

    numeric: dict[str, dict] = {}
    categorical: dict[str, dict] = {}
    col = 0
    for _, trans, cols in _blocks(pre):
        tname = type(trans).__name__
        if tname == "SplineTransformer":
            per = trans.transform(np.zeros((1, len(cols)))).shape[1] // len(cols)
            for j, feat in enumerate(cols):
                block = clf.coef_[:, col + j * per: col + (j + 1) * per]
                numeric[feat] = _spline_segments(trans, j, block)
            col += per * len(cols)
        elif tname == "StandardScaler":
            for j, feat in enumerate(cols):
                w = clf.coef_[:, col + j]
                numeric[feat] = {"kind": "linear", "mean": float(trans.mean_[j]),
                                 "scale": float(trans.scale_[j]), "w": [float(x) for x in w]}
            col += len(cols)
        elif tname == "OneHotEncoder":
            for j, feat in enumerate(cols):
                cats = [str(c) for c in trans.categories_[j]]
                block = clf.coef_[:, col: col + len(cats)]
                categorical[feat] = {c: [float(block[k, i]) for k in range(n_eff)]
                                     for i, c in enumerate(cats)}
                categorical[feat]["_unknown"] = [0.0] * n_eff
                col += len(cats)
        else:  # pragma: no cover
            raise ValueError(f"unhandled transformer {tname}")

    return {
        "format": "linear-portable-1",
        "objective": objective,
        "classes": classes,
        "labels": list(labels),
        "feature_names": list(feature_names),
        "intercept": [float(b) for b in clf.intercept_],
        "numeric": numeric,
        "categorical": categorical,
    }


# ---- reference evaluator (spec for the TS port) --------------------------

def _numeric_contrib(entry: dict, x: float, k: int) -> float:
    if entry["kind"] == "linear":
        return ((x - entry["mean"]) / entry["scale"]) * entry["w"][k]
    breaks, seg = entry["breaks"], entry["seg"]
    if x <= breaks[0]:
        s, u = 0, 0.0
    elif x >= breaks[-1]:
        s, u = len(seg) - 1, breaks[-1] - breaks[-2]
    else:
        s = 0
        while s + 1 < len(breaks) - 1 and x >= breaks[s + 1]:
            s += 1
        u = x - breaks[s]
    a, b, c, d = seg[s][k]
    return a + b * u + c * u * u + d * u * u * u


def predict_proba_linear(model: dict, x: dict) -> dict[str, float]:
    n_eff = len(model["intercept"])
    dec = list(model["intercept"])
    for feat, entry in model["numeric"].items():
        v = x.get(feat)
        if v is None or (isinstance(v, float) and math.isnan(v)):
            continue
        for k in range(n_eff):
            dec[k] += _numeric_contrib(entry, float(v), k)
    for feat, table in model["categorical"].items():
        row = table.get(str(x.get(feat)), table["_unknown"])
        for k in range(n_eff):
            dec[k] += row[k]

    classes, labels = model["classes"], model["labels"]
    if model["objective"] == "binary":
        p1 = 1.0 / (1.0 + math.exp(-dec[0]))
        by = {classes[0]: 1.0 - p1, classes[1]: p1}
    else:
        m = max(dec)
        ex = [math.exp(d - m) for d in dec]
        s = sum(ex)
        by = dict(zip(classes, (e / s for e in ex)))
    return {l: by[l] for l in labels}
